lib/fizzy/stack.hpp

```diff
@@ -35,4 +37,65 @@ class Stack : public std::vector<T>
     /// Drops @a num_elements elements from the top of the stack.
     void drop(size_t num_elements = 1) noexcept { resize(size() - num_elements); }
 };
+
+class OperandStack
+{
+    /// The pointer to the top item, or below the stack bottom if stack is empty.
+    uint64_t* m_top;
+
+    /// The storage for items.
+    std::unique_ptr<uint64_t[]> m_storage;
+
+public:
+    /// Default constructor. Sets the top item pointer to below the stack bottom.
+    explicit OperandStack(size_t max_stack_height)
+      : m_top{nullptr} /* temporarily */, m_storage{new uint64_t[max_stack_height]}
+    {
+        shrink(0);
+    }
+
+    OperandStack(const OperandStack&) = delete;
+    OperandStack& operator=(const OperandStack&) = delete;
+
+    /// The current number of items on the stack (aka stack height).
+    size_t size() noexcept { return static_cast<size_t>(m_top + 1 - &m_storage[0]); }
+
+    /// Returns the reference to the top item.
+    /// Requires non-empty stack.
+    auto& top() noexcept
+    {
+        assert(size() != 0);
+        return *m_top;
+    }
+
+    /// Returns the reference to the stack item on given position from the stack top.
+    /// Requires index < size().
+    auto& operator[](size_t index) noexcept
+    {
+        assert(index < size());
+        return *(m_top - index);
+    }
+
+    /// Pushes an item on the stack.
+    /// The stack max height limit is not checked.
```

Comment:
You could add assert to check this.

Reply:
Hard to do without keeping `max_stack_height`.